### DISCUSSION TITLE
Template Engine e Arquitetura SSR

### DISCUSSION THREAD

**Discussion by aronreisx** [frontendbr/forum#2018] [Discussão]
Olá, pessoal.
Estou estudando desenvolvimentno web a alguns meses e tenho procurado evoluir mais no que diz respeito a arquitetura das aplicações.
Estive pensando sobre Arquitetura Server-Side Rendering com Template Engine, mais especificamente com Handlebars. Fiz um diagrama arquitetural e queria que me dissesem o que  acham:

<a href="https://ibb.co/CtkXSf5"><img src="https://i.ibb.co/kyRrPtg/ssr-hanndlebars.jpg" alt="ssr-hanndlebars" border="0"></a>

Qualquer sugestão é válida. O intuito do post é trocar idéias mesmo.
Obrigado a quem puder ajudar.

**Comment by wilsonneto-swe** [comment]
Se for algo para produção eu retiraria o SQLite por um Postgres ou outro banco mais confiável e performático. Sobre handlebars, eu partiria para uma arquitetura desacoplada, onde seu front é idependente como um next + react, ou vue + nuxt, consumindo de uma aplicação API.

![image](https://user-images.githubusercontent.com/20674439/107866382-19d82700-6e4f-11eb-9a98-3c40a8691100.png)

**Comment by ninetails** [comment]
![draft de diagrama de SSR](https://user-images.githubusercontent.com/40065/107992374-532ba680-6fb7-11eb-9a3d-ba7d16468135.png)

Só dizendo que em relação ao Handlebars, que eu me lembre eu não tinha conseguido na época utilizar a mesma aplicação tanto pro backend quanto pro frontend.

E os adapters são opcionais, mas eu os adicionei no exemplo pq é bom que tua aplicação seja independente de qual Template Engine você esteja utilizando. Talvez também precise de alguns mapeadores. Também, claro, tendo provável reescrita de templates mas você poderia por exemplo trocar de Handlebars para Pug (antigo Jade) caso teu código for abstraído o bastante para tal no caso de ter algum adaptador no meio do caminho.

Importante é que tua aplicação não tenha que lidar com o banco de dados e possivelmente também não com autenticação e autorização (exceto somente transporte, como no caso do front que vai precisar de tela de login e trafegar se o usuário está logado) pois abriria brechas para falhas de segurança. Só abriria uma exceção para serviços como **Firebase** que apesar de ter um "Banco de dados" para vc utilizar, é possível configurar quais dados o front e o usuário possam acessar.

E claro, considere isto como um _draft_ à discussão. Não existe melhor arquitetura, só estou jogando uma proposta.
